enterprise/coderd/usage/publisher.go

```diff
@@ -0,0 +1,427 @@
+// validation.
+func PublisherWithLicenseKeys(keys map[string]ed25519.PublicKey) TallymanPublisherOption {
+	return func(p *tallymanPublisher) {
+		p.licenseKeys = keys
+	}
+}
+
+// PublisherWithIngestURL sets the ingest URL to use for publishing usage
+// events.
+func PublisherWithIngestURL(ingestURL string) TallymanPublisherOption {
+	return func(p *tallymanPublisher) {
+		p.ingestURL = ingestURL
+	}
+}
+
+// PublisherWithInitialDelay sets the initial delay for the publisher.
+func PublisherWithInitialDelay(initialDelay time.Duration) TallymanPublisherOption {
+	return func(p *tallymanPublisher) {
+		p.initialDelay = initialDelay
+	}
+}
+
+// Start implements Publisher.
+func (p *tallymanPublisher) Start() error {
+	deploymentID, err := p.db.GetDeploymentID(p.ctx)
+	if err != nil {
+		return xerrors.Errorf("get deployment ID: %w", err)
+	}
+	deploymentUUID, err := uuid.Parse(deploymentID)
+	if err != nil {
+		return xerrors.Errorf("parse deployment ID %q: %w", deploymentID, err)
+	}
+
+	if p.initialDelay <= 0 {
+		// Pick a random time between tallymanPublishInitialMinimumDelay and
+		// tallymanPublishInterval.
+		maxPlusDelay := int(tallymanPublishInterval - tallymanPublishInitialMinimumDelay)
+		plusDelay, err := cryptorand.Intn(maxPlusDelay)
+		if err != nil {
+			return xerrors.Errorf("could not generate random start delay: %w", err)
+		}
+		p.initialDelay = tallymanPublishInitialMinimumDelay + time.Duration(plusDelay)
+	}
+
+	go p.publishLoop(p.ctx, deploymentUUID)
+	return nil
+}
+
+func (p *tallymanPublisher) publishLoop(ctx context.Context, deploymentID uuid.UUID) {
+	defer close(p.done)
+
+	// Start the ticker with the initial delay. We will reset it to the interval
+	// after the first tick.
+	ticker := p.clock.NewTicker(p.initialDelay)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+
+		err := p.publish(ctx, deploymentID)
+		if err != nil {
+			p.log.Warn(ctx, "publish usage events to tallyman", slog.Error(err))
+		}
+		ticker.Reset(tallymanPublishInterval)
+	}
+}
+
+// publish publishes usage events to Tallyman in a loop until there is an error
+// (or any rejection) or there are no more events to publish.
+func (p *tallymanPublisher) publish(ctx context.Context, deploymentID uuid.UUID) error {
+	for {
+		publishCtx, publishCtxCancel := context.WithTimeout(ctx, tallymanPublishTimeout)
+		accepted, err := p.publishOnce(publishCtx, deploymentID)
+		publishCtxCancel()
+		if err != nil {
+			return xerrors.Errorf("publish usage events to tallyman: %w", err)
+		}
+		if accepted < tallymanPublishBatchSize {
+			// We published less than the batch size, so we're done.
+			return nil
+		}
+	}
+}
+
+// publishOnce publishes up to tallymanPublishBatchSize usage events to
+// tallyman. It returns the number of successfully published events.
+func (p *tallymanPublisher) publishOnce(ctx context.Context, deploymentID uuid.UUID) (int, error) {
+	licenseJwt, err := p.getBestLicenseJWT(ctx)
```

Comment:
We shouldn't be doing this for every publish. Can we connect it up to the rest of the license code so we don't add more queries? Or at least use the same query strategy of querying every 10 minutes and listening for published changes?

Reply:
> We shouldn't be doing this for every publish.

This is very light so I don't see the issue with querying it every time. The reason why the entitlements check isn't performed on demand and is cached is that it's used on almost every single request to the API. In general, this is being done a few times every 17 minutes per replica. The licenses table is also very small.

> Can we connect it up to the rest of the license code so we don't add more queries?

The license code doesn't actually return much information about licenses themselves, or perform any license selection that matches our requirements. Quite frankly it's in desperate need of a rewrite, but it's out of scope for this PR.

Reply:
You're probably right that it's a very lightweight query, so I'm fine with this.

I do agree the license code is in need of a refactor, which could hopefully reduce some code complexity here, even if it's negligble perf impact. But, I also agree that's out of scope.